### PR TITLE
[SPARK-19956][Core]Optimize a location order of blocks with topology information

### PR DESCRIPTION
**Pull request by ConeyLiu**
## What changes were proposed in this pull request?

When call the method getLocations of BlockManager, we only compare the data block host. Random selection for non-local data blocks, this may cause the selected data block to be in a different rack. So in this patch to increase the sort of the rack.

## How was this patch tested?

New test case.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by jerryshao** [comment]
The fix LGTM, I think it is nice to have such topology priority. CC @mridulm .

File: core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala [hunk 1]
```diff
@@ -500,6 +500,30 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
...
+    val bmId4 = BlockManagerId("id4", otherHost, 4, Some(otherRack))
+    val bmId5 = BlockManagerId("id5", otherHost, 5, Some(localRack))
+    when(bmMaster.getLocations(mc.any[BlockId]))
+      .thenReturn(Seq(bmId1, bmId2, bmId3, bmId3, bmId4, bmId5))
+
+    val blockManager = makeBlockManager(128, "exec", bmMaster)
+    val getLocations = PrivateMethod[Seq[BlockManagerId]]('getLocations)
+    val locations = blockManager invokePrivate getLocations(BroadcastBlockId(0))
+    assert(locations.map(_.host).toSet
+      === Set(localHost, localHost, otherHost, otherHost, otherHost))
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala]: Remove `toSet` and make `Set` as `Seq` ?
Making it a Set sort of looses the point of this change (ordering is lost).
Same for topology info below below.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 2]
```diff
@@ -555,12 +555,15 @@ private[spark] class BlockManager(
 
   /**
    * Return a list of locations for the given block, prioritizing the local machine since
-   * multiple block managers can share the same host.
+   * multiple block managers can share the same host, then try to get the same rack data.
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: How about "then try to get the same rack data" -> "followed by hosts on the same rack" ?

**Comment by ConeyLiu** [comment]
Hi, @jerryshao @mridulm Thanks for your review, I have updated the code.

File: core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala [hunk 3]
```diff
@@ -497,7 +497,30 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
     val blockManager = makeBlockManager(128, "exec", bmMaster)
     val getLocations = PrivateMethod[Seq[BlockManagerId]]('getLocations)
     val locations = blockManager invokePrivate getLocations(BroadcastBlockId(0))
-    assert(locations.map(_.host).toSet === Set(localHost, localHost, otherHost))
+    assert(locations.map(_.host) === Seq(localHost, localHost, otherHost))
```
Comment (ConeyLiu) [inline core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala]: There should also be Seq.

**Review by mridulm — APPROVED** [review]
Looks good to me, will keep it open in case there are other comments.
+CC @jerryshao , @ConeyLiu

**Comment by ConeyLiu** [comment]
Ok, thanks a lot.

**Comment by ConeyLiu** [comment]
Hi, @cloud-fan @zsxwing Can you take a look? Thanks a lot.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by ConeyLiu** [comment]
retest this please

**Comment by mridulm** [comment]
Will merge when tests pass.

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 4]
```diff
@@ -555,12 +555,15 @@ private[spark] class BlockManager(
...
    * Return a list of locations for the given block, prioritizing the local machine since
-   * multiple block managers can share the same host.
+   * multiple block managers can share the same host, followed by hosts on the same rack.
    */
   private def getLocations(blockId: BlockId): Seq[BlockManagerId] = {
     val locs = Random.shuffle(master.getLocations(blockId))
     val (preferredLocs, otherLocs) = locs.partition { loc => blockManagerId.host == loc.host }
-    preferredLocs ++ otherLocs
+    val (sameRackLocs, differentRackLocs) = otherLocs.partition {
+      loc => blockManagerId.topologyInfo == loc.topologyInfo
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: If `blockManagerId.topologyInfo` is `None`, we will prefer the locations with empty `topologyInfo`. It is slightly different with what the shuffling wants to do here.

Comment (ConeyLiu) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Modified, thanks a lot for the good advice.

**Comment by ConeyLiu** [comment]
Thanks both of you for review, I have addressed the comments and modified the test case. Please help calling jenkins for test, because I can't trigger that.  Thanks again.

**Comment by mridulm** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by mridulm** [comment]
That is weird, jenkins should have restarted build when I commented ...

**Comment by gatorsmile** [comment]
retest this please

**Comment by viirya** [comment]
LGTM

**Comment by ConeyLiu** [comment]
Thanks for your review.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by mridulm** [comment]
Thanks for merging @cloud-fan, this PR kept dropping form my list ...

**Comment by ConeyLiu** [comment]
Thanks @cloud-fan @mridulm @gatorsmile

**Pull request closed without merging**